### PR TITLE
Self curse implementation and changes to curses from config

### LINKED ISSUE #5056: Reservation mastery calculation bug
The reservation mastery "8% increased Damage for each of your Aura or Herald Skills affecting you" is currently bugged and calculating a DPS increase for curse-based auras which do not 'affect you'.

See my post on the officlal PoE forums here with my in-game tests: 
https://www.pathofexile.com/forum/view-thread/3308443

This causes a significant difference in the POB DPS from that mastery and the actual in-game DPS (it is half as effective as POB shows for my current build)

### PR DESCRIPTION
**Pull request by Paliak**
Fixes #5056 , https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/5069.

### Description of the problem being solved:
This pr improves ui and curse mechanics. Implements self curse handling and fixes related issues.

Currently work in progress.
TODO:
- Drop down for curses in the config. :heavy_check_mark:
- Handling for increased curse effect from curses. :heavy_check_mark:
- Handling multiple sources of the same curse with different strengths  :heavy_check_mark:
- Make config curses take priority.  :heavy_check_mark:
- Implement handling for reflected curses.  :heavy_check_mark:
- Implement self duration modifiers from temp chains and other sources  :heavy_check_mark:
- Refactor the drop down list selection into a new control ?

### CODE REVIEW

**Comment by Nightblade** [comment]
Would you mind using "no" (or similar) instead of "dont" in variable names?  Sorry for the pedantry.  :)

**Comment by QuickStick123** [comment]
Less curse effect against players on gem doesn't seem to work but does for config?
Maybe add self curses to the Aura and Buff Skills config section so you can see them apply effects to the player. 
image
Divine blessing aura doesn't seem to work temp chains reduced buff effect. There are some skills that it probably shouldn't effect e.g. convocation/galvanic field.
These could probably use spaces
image
Hexproof is currently implemented as 100% reduced curse effect on self. This might stop reflected marks from applying to the player. https://www.poewiki.net/wiki/Hexproof

**Comment by QuickStick123** [comment]
image
Phantasmal haste might have slightly different functionallity to temp chains and might stack additivly instead of multiplicibly. (faster bleed is inc according to PoB testing might be a good idea.) But is implemeneted the same way.

**Comment by Paliak** [comment]
>Was the fixes meant to be https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/5069 or it fixes both?

It was initially meant to fix https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/5069 but ever since @Regisle mentioned he'd like to implement some changes this pr blew up in scope as i read the code more and had more and more ideas. I'll likely focus on breaking this pr up into 2 separate ones:

- Fix for: https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/5056 and Ui changes for config curses as the fix is rather simple and those two were the main things i had in mind with this pr
- Temp chains duration mod as a separate pr as it will require more changes than i initially expected.

**Comment by Paliak** [comment]
Switching back to draft in the mean time as i move things around.

**Comment by QuickStick123** [comment]
Hex proof is currently implemented as 100% reduced effect of curses. This will still apply to marks. A new mod probably needs implementing hexEffectOnSelf as 100% less or a mod flag such as I think if you are an occulist and use can apply to hexproof enimies you might be able to override vidiri's veil.

**Comment by QuickStick123** [comment]
From my testing it seems occultist doesn't work on players. Which makes sense as it applies to enemies atleast with shackles.
Rotblood applies regardless of hex proof state.
Marks don't work with shackles / rotblood

**Comment by Paliak** [comment]
I left the player mark slot code in place as i think it may be useful to let the `ExtraCurse` apply them. Even though it doesn't do anything currently.

**Comment by Paliak** [comment]
Support for duration of reflected curses required https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/5102 to be merge back in here.

**Comment by QuickStick123** [comment]
image
https://pobb.in/n8UnEKeUNEg1
Goto grace socket group.

**Comment by deathbeam** [comment]
Problems I noticed while testing this:
1. Its increasing duration of stuff like frost shield and spark what is incorrect (but I think this was already mentioned earlier)
2. Self curse effect mods are also applying to the enemy part of the curse, e.g suddenly I am getting a lot of damage from self curse mods with Temp Chains on poison build and that is also incorrect

**Comment by Paliak** [comment]
Spark being affected is fixed. The fix for frost shield might cause some skills that should be affected to not be affected. Can you send me the build for the poison issue @deathbeam ? I can't seem to really replicate.

**Comment by deathbeam** [comment]
Hmm when I pulled your last 2 commits the issue dissapeared I think (or at least pob is showing more reasonable damage) but go to commit 87ab9dead97988a30d910ba5b8186e02f44d7dac with this https://pobb.in/L4o78bt1vMUU and check poison duration.

**Comment by deathbeam** [comment]
Actually nvm in your latest change temp chains just stopped working completely for poison duration

**Comment by Paliak** [comment]
Matching buffs/debuffs correctly is quite the challenge. Temp chains doesn't seem to affect poison duration in release pob so i assume it's a separate issue. I'll look into it.

**Comment by deathbeam** [comment]
Temp chains definitely affects poison duration in release pob

**Comment by Paliak** [comment]
Yeah just noticed i had it setup wrong. `durationMult` is reused for ailment duration and that seems to cause issues. Working on a solution.

**Comment by QuickStick123** [comment]
https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/5102#issuecomment-1256842896
Most of the unticked ones are either not supported durations or debuffs that regular temp chains should apply too (not self curse.)

**Comment by QuickStick123** [comment]
Moving this here for clarity.
- [x] blood rage
- [x] winter orb
- [ ] bv
- [x] cf (buff)
- [ ] self curses except temp chains
- [ ] link skills
- [x] immortal call,  steel skin
- [ ] molten shell
- [ ] warcries
- [ ] static strike (buff)
- [x] smite (buff)
- [ ] boneshatter (trauma)
- [ ] vigilant strike (fortify)
- [x] all buff/debuffs (charges/fortify/onsalught/poision/ignited etc).
- [x] divine blessing
- [ ] lifetap / behead / cruelty / inspiration
- [ ] cyclone
- [ ] smoke mine (secondary buff)
- [ ] venom gyre blades buff
- [ ] flicker strike movespeed buff
- [ ] shattering steel (steel ward buff)
- [ ] ambush (buff only skill duration secondary is a debuff enemies expiry will work.)
- [x] phase run (primary obviously as phasing secondary not sure.)

no

- [ ] herald of purity
- [x] bladestorm
- [x] conc path
- [ ] dom blow minion duration (primary duration may be a debuff applied to enemies and should scale with enemy expiry rate same with other similar mechanics e.g. absolution / herald of purity)
- [x] all other minions 
- [x] all other attack skills e.g. earthquake, rage vortex, bladestorm, etc.
- [x] all totems
- [ ] all banners
- [x] all mines / traps
- [ ] artic armour 
- [x] frost blink / flame dash / creeping frost / vortex / cold snap /(the duration is the trail of the chill ground)
- [ ] Frost bomb (secondary is a debuff and should scale with enemy expiry).
- [x] ek / bladefall 
- [x] desecrate + alchemist mark
- [x] mirror arrow + blink arrow
- [ ] tornado primary is a debuff on enmies hinder secondary is length of tornado (unscalable by expiry rate)
- [ ] perstillent strike (debuffs on enmies)
- [x] cremation
- [ ] withering step / withering touch / wither (wither debuff will scale with enemy expiry rate)
- [x] frost wall
- [ ] blight (debuff)
- [ ] siphoning trap (debuff)
- [x] lightning spire / seismic / flamethrower / explosive (created effect that occurs for time period)
- [x] storm call \ flame surge (created thing duraiton)
- [x] offerings (don't seem to work didn't test with necro node.)
- [ ] galvanic field
- [x] frost shield / sigil of power
- [x] minion ability durations.
- [x] hydrosphere + void sphere.

unsure
stun duration?
leech?
convokation (there is no buff so I suspect this doesn't work but is impossible to test as you can't get enough following minions. maybe with max zombies + alberon + skelly warriors.)
herald of thunder

**Comment by Paliak** [comment]
Marking for review again. I tested the skills [QuickStick123](https://github.com/QuickStick123) posted above and they all seem to work correctly. I could not get the drop down to break either. Someone should still look it over before merging though.

**Comment by QuickStick123** [comment]
This hasn't updated it just checked over everything. A lot of these durations just aren't classified or implemented. The only real major ones that I would like fixed are:
- galvanic field
- blade vortex
- arctic armour
- molten shell
- link skills
- herald of purity
- warcries
The rest will see minimal use and most already aren't supported.

**Comment by Paliak** [comment]
Galvanic doesn't work.
Blade vortex doesn't.  I think i misunderstood your list. 

Arctic seems to work 
obraz

Molten shell doesn't work.

Links don't work.

Herald of purity seems to work
obraz

War cries don't work.

Some Banners seem to work
obraz


I'll implement the ones you listed and look into adding the duration tags where needed.

I'm also considering reworking the drop down menu into a separate class.

**Comment by QuickStick123** [comment]
Unticked ones behave differently ingame is basicly how I structured my list.

**Comment by Paliak** [comment]
Oh, Okay. I'll test again and get to fixing. Thanks for checking.

**Comment by QuickStick123** [comment]
Here is a PoB with all the skills I was testing so you don't have to manually add them.
https://pobb.in/3u0Cd16ZDudK

**Comment by deathbeam** [comment]
arctic armour shouldnt work, the duration on the skill is duration of the chilled ground pretty sure, it shouldnt be affected by expiry rate. same with durations of minions. Banners shouldnt work either, at leats not the base duration, its duration of the banner on ground.

**Comment by Paliak** [comment]
Yeah i misunderstood QuickStick's list. Working on a fix.

**Comment by QuickStick123** [comment]
There are still some expiry rate mod using more multipliers instead of inc.

**Comment by QuickStick123** [comment]
Galvanic field probably shouldn't scale with self curse.
Cyclone duration should scale with stages.
Sometimes breakdown seems to show mod even though it doesn't apply
image
The missing durations might be caused by https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/5125.

**Comment by Paliak** [comment]
Galvanic Field doesn't seem to be affected by self curse temp chains. On a fresh build.
obraz

The Cyclone duration is the duration of each stage. Should i make it show total duration instead ? 

Can't seem to replicate the issue with the breakdown showing the mod on a fresh build. It is broken on the build  you linked though which is interesting. Also the entire config page seems mangled after importing the build you linked.
obraz

Not sure what's going on there. I'll check the build xml.

**Comment by QuickStick123** [comment]
Most of the other staged duration skills e.g. winter orb scale their duration per stage with stages so yes for cyclone.
I can replicate galvanic field misbehaving on fresh build.
https://pobb.in/c4_CaggrTTIv
image

The tooltip for blight misappearing seems to be caused by enduring cry.
image
image
https://pobb.in/HMFUW_jfMvc-

**Comment by QuickStick123** [comment]
Everything looks good expect maybe shade form (replica eternity shroud) not working and debuffs not working correctly with regular temp chains (exists in master). I would mark this ready for review.

**Comment by Paliak** [comment]
I referenced the wiki for the expiry rate modifiers.
 I'd need more details to fix the normal temp chains interaction.
Breakdowns showing a mod that does not apply may still be the case for the secondary durations. I'm a bit hesitant to add yet another set of flags though. I'll try to think of a better solution.

**Comment by QuickStick123** [comment]
image

**Comment by Paliak** [comment]
> [The modifier expire x% faster has diminishing returns, while the modifier expire x% slower has increasing returns.](https://www.poewiki.net/wiki/Expiration_rate)

>[Effects on you and nearby allies expire slower.](https://www.poewiki.net/wiki/Shaper%27s_Presence)

I guess i'm taking this to literally. 

>There are definitely mods that apply exclusively to buff/debuff instead of generic I haven't checked if these behave correctly. e.g. phantasmal haste / divergent temporal rift / duration debuff expiry rate mastery
/ torrents reclimation / warped time piece / haste watchers / statis prison / pantheon.

Those should work correctly.
obraz

File: src/Modules/CalcOffence.lua [hunk 1]
```diff
@@ -276,18 +276,79 @@ local function calcWarcryCastTime(skillModList, skillCfg, actor)
 	return warcryCastTime
 end
 
-function calcSkillDuration(skillModList, skillCfg, skillData, env, enemyDB)
-	local durationMod = calcLib.mod(skillModList, skillCfg, "Duration", "PrimaryDuration", "SkillAndDamagingAilmentDuration", skillData.mineDurationAppliesToSkill and "MineDuration" or nil)
-	local durationBase = (skillData.duration or 0) + skillModList:Sum("BASE", skillCfg, "Duration", "PrimaryDuration")
-	local duration = durationBase * durationMod
-	local debuffDurationMult = 1
+local function getDurationMult(skill, env, enemyDB, isAilement)
```
Comment (QuickStick123) [inline src/Modules/CalcOffence.lua:279]:
```suggestion
local function getDurationMult(skill, env, enemyDB, isAilment)
```

File: src/Modules/ModParser.lua [hunk 2]
```diff
@@ -2814,7 +2818,7 @@ local specialModList = {
 	["left ring slot: cover enemies in ash for 5 seconds when you ignite them"] = { mod("CoveredInAshEffect", "BASE", 20, { type = "SlotNumber", num = 1 }, { type = "ActorCondition", actor = "enemy", var = "Ignited" }) },
 	["right ring slot: cover enemies in frost for 5 seconds when you freeze them"] = { mod("CoveredInFrostEffect", "BASE", 20, { type = "SlotNumber", num = 2 }, { type = "ActorCondition", actor = "enemy", var = "Frozen" }) },
 	["([%a%s]+) has (%d+)%% increased effect"] = function(_, skill, num) return { mod("BuffEffect", "INC", num, { type = "SkillId", skillId = gemIdLookup[skill]}) } end,
-	["debuffs on you expire (%d+)%% faster"] = function(num) return { mod("SelfDebuffExpirationRate", "BASE", num)} end,
+	["debuffs on you expire (%d+)%% faster"] = function(num) return {mod("EffectExpiresFaster", "INC", nil, 0, 0, {type = "SkillCondition", debuff = true}),} end,
```
Comment (QuickStick123) [inline src/Modules/ModParser.lua:2821]:
```suggestion
	["debuffs on you expire (%d+)%% faster"] = function(num) return {mod("EffectExpiresFaster", "INC", num, 0, 0, {type = "SkillCondition", debuff = true}),} end,
```

**Comment by Paliak** [comment]
Tests fail because ModCache is not up to date.

File: src/Modules/CalcPerform.lua [hunk 3]
```diff
@@ -2133,49 +1947,59 @@ function calcs.perform(env, avoidCache)
...
 				})
-				local curseModList = { }
+
+				local modScale = nil
+				if value.applyToPlayer then
+					local cfg = { skillName = grantedEffect.name }
+					local inc = modDB:Sum("INC", cfg, "CurseEffectOnSelf") + gemModList:Sum("INC", nil, "CurseEffectAgainstPlayer")
+					local more = modDB:More(cfg, "CurseEffectOnSelf") * gemModList:More(nil, "CurseEffectAgainstPlayer")
+					local configInc = value.effect and (1 + value.effect / 100) or 1
+					modScale = (1 + inc / 100) * more * configInc
```
Comment (QuickStick123) [inline src/Modules/CalcPerform.lua:1957]:
```suggestion
					modScale = (1 + inc / 100) * more * configInc
```
Does map curse effect act as more multiplier or does it stack additively with regular increased curse effect on player.

**Comment by QuickStick123** [comment]
Supreme ego should still allow self curse with rotblood promise
https://pobb.in/TUKhwfe3Om58

**Comment by QuickStick123** [comment]
Boneshatter trauma duration code is in probably ensure the expiry rate works correctly with it.

File: src/Modules/CalcPerform.lua (hunk 3 above)
Comment (Paliak) [reply src/Modules/CalcPerform.lua]: The effect box has been removed so that value is not longer needed.

File: src/Modules/ConfigOptions.lua [hunk 4]
```diff
@@ -566,45 +566,76 @@ Huge sets the radius to 11.
...
+		if build.configTab.varControls['playerCursedWithTemporalChains'].shown then
+			modList:NewMod("ExtraCurse", "LIST", { skillId = "TemporalChains", level = val, applyToPlayer = true, configCurse = true })
+		end
 	end },
-	{ var = "playerCursedWithVulnerability", type = "count", label = "Vulnerability:", tooltip = "Sets the level of Vulnerability to apply to the player.\nIn mid tier maps, 'of Vulnerability' applies level 10.\nIn high tier maps, 'of Vulnerability' applies level 15.", apply = function(val, modList, enemyModList)
-		modList:NewMod("ExtraCurse", "LIST", { skillId = "Vulnerability", level = val, applyToPlayer = true })
+	{ var = "playerCursedWithVulnerability", type = "count", label = "Vulnerability:", tooltip = "Sets the level of Vulnerability to apply to the player.\nIn mid tier maps, 'of Vulnerability' applies level 10.\nIn high tier maps, 'of Vulnerability' applies level 15.", saveShown = true, defaultPlaceholderState = 10, defaultHidden = true, removeBox = true, apply = function(val, modList, enemyModList, build)
```
Comment (QuickStick123) [inline src/Modules/ConfigOptions.lua:630]:
```suggestion
	{ var = "playerCursedWithVulnerability", type = "count", label = "Vulnerability:", tooltip = "Sets the level of Vulnerability to apply to the player.\nIn mid tier maps, 'of Vulnerability' applies level 10.\nIn high tier maps, 'of Vulnerability' applies level 15.", saveShown = true, defaultPlaceholderState = 10, defaultHidden = true, removeBox = true, apply = function(val, modList, enemyModList, build)
```
These tooltips are outdated see https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues/5560

File: src/Classes/ConfigTab.lua [hunk 5]
```diff
@@ -265,6 +279,26 @@ local ConfigTabClass = newClass("ConfigTab", "UndoHandler", "ControlHost", "Cont
 					end
 					return innerLabel
 				end
+				local innerTooltipFunc = control.tooltipFunc
```
Comment (deathbeam) [inline src/Classes/ConfigTab.lua:282]: looks like merge/rebase went wrong, tooltipFunc logic is here twice

Comment (Paliak) [reply src/Classes/ConfigTab.lua]: Thanks for checking. Should be fixed by [2070ff3](https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/5074/commits/2070ff37d0a5035926a9542f45dcd48110fc60cc)

**Comment by Paliak** [comment]
I've added support for the `Your Elemental Resistances cannot be lowered by Curses` mod here instead of a new pr due to this pr introducing better on player curse handling.

**Comment by Paliak** [comment]
`base_tertiary_skill_effect_duration` has been added since this pr was written and is not supported. Likely good chunks of this pr would need to be updated to work correctly.

**Comment by Paliak** [comment]
Ui code superseded by <https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/8203>. The self curse parts need a rework and it seems @Regisle is working on that so this should be good to close.

**Pull request closed without merging**